### PR TITLE
Added filter by owner for challenges

### LINKED ISSUE #1809: Add "Created" or "Owner" option to Challenge list filters
The "Participating" / "Not Participating" options for filtering the Challenge list are quite useful, as are the group filters (especially with #1781 accomplished, yay!). One filter I found myself wanting but lacking, though, was a filter to show only the Challenges I created or am responsible for.

It doesn't fit well as an additional option on the other filter lists... it'd probably work something like the Participating filter, "Owned" / "Not Owned" / "Either"?

### PR DESCRIPTION
**Pull request by TheHollidayInn**
Fixes https://github.com/HabitRPG/habitrpg/issues/1809

Reboot for: https://github.com/HabitRPG/habitrpg/pull/4842

### CODE REVIEW

File: website/views/options/social/challenges.jade [hunk 1]
```diff
@@ -91,6 +91,19 @@ script(type='text/ng-template', id='partials/options.social.challenges.html')
...
+            label
+              input(type='radio', name='search-participation-radio', ng-click='search._isOwner = true')
+              =env.t('challengedOwnedFilter')
+          .radio
+            label
+              input(type='radio', name='search-participation-radio', ng-click='search._isOwner = false')
+              =env.t('challengedNotOwnedFilter')
+          .radio
+            label
+              input(type='radio', name='search-participation-radio', ng-click='search._isOwner = undefined')
```
Comment (crookedneighbor) [inline website/views/options/social/challenges.jade]: The ownership radio buttons should default to "Either" like Membership does:

screen shot 2015-04-11 at 4 21 06 pm

Comment (crookedneighbor) [inline website/views/options/social/challenges.jade]: Oh, I see, you can't filter by membership AND ownership. I guess that makes sense.

Comment (deilann) [inline website/views/options/social/challenges.jade]: Why? You can own and not be a member of a challenge.

**Comment by crookedneighbor** [comment]
I can't edit challenges with this logic in there. The edit form just never appears.

**Comment by deilann** [comment]
But why does that make sense?

**Comment by TheHollidayInn** [comment]
@crookedneighbor Can you help me with rebasing. When I try to rebase with origin/[branch name] and then push, I get all the new commits added to my branch as shown above.

Also, I have created a fix for the editing and defaulted the option to either.

@deilann I have set the options so that you can select a membership filter and ownership filter at the same time. However, if you have created the challenge (i.e., are the owner), by the logic currently in plac you are considered participating. Here is a picture to show:

screen shot 2015-04-13 at 11 28 33 am

There are 0 participants, but I am still considered participating.

**Comment by crookedneighbor** [comment]
@TheHollidayInn I noticed that, if you add a second person, the number jumps to 2.

**Comment by crookedneighbor** [comment]
@TheHollidayInn Re: rebasing, I've cherry-picked the two relevant commits and put them in an upstream branch. Try this (assuming you have the upstream remote set up):

```
git reset --hard upstream/theHollidayInn-Git_Rebase_of_Doom
git push --force
```

This should reset your current branch to the version I've made on upstream and then overwrite your remote branch for the PR.

**Comment by deilann** [comment]
Yes, if no one joins a challenge, it shows 0 participants. However, you can
still leave that challenge!

This is incredibly useful sometimes. There are challenges you want to be
able to control, but not be a part of.

I'm just not sure _why_ it makes sense that it should be this way. In fact,
the presumed membership may make this not as useful for me.

**Comment by TheHollidayInn** [comment]
@crookedneighbor Okay, will do that. I'm going to reread your past comment to try and fix my workflow, and if that doesn't work, I'll show you want I am doing.

@deilann Yes, I see that functionality happening. I can join the filters together with an "and" or "or" logic if need be, but since being an owner messes with the participating filter, everything might not display as expected.

Should I also fix the "being counted as a participant when an owner" functionality?

**Comment by deilann** [comment]
I know that when I leave a challenge I own, it does not show up on my
participating filter with the way the filters currently work. So I'm not
sure what you mean by "being an owner messes with the participating filter"

**Comment by TheHollidayInn** [comment]
Well, that works. I didn't know if you were saying that behavior was intentional.

**Comment by deilann** [comment]
That's what I was asking @crookedneighbor.

"Oh, I see, you can't filter by membership AND ownership. I guess that
makes sense."

I don't see why that makes sense.

**Comment by crookedneighbor** [comment]
@TheHollidayInn Have you had a chance to try out the git commands I sent? I noticed the rogue commits are still here, so either you haven't or it didn't work. If it didn't work, let me know!

FYI, the branch I made for you should be up to date with develop. So no additional rebasing will be needed for this PR.

**Comment by TheHollidayInn** [comment]
Hey, I just tested it out. Here is what I got:

```
Keiths-MacBook-Pro:habitrpg krh121791$ git reset --hard upstream/theHollidayInn-Git_Rebase_of_Doom
fatal: ambiguous argument 'upstream/theHollidayInn-Git_Rebase_of_Doom': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

I have upstream setup, I believe, because I rebase with upstream/develop all the time.

**Comment by TheHollidayInn** [comment]
I think I am having this issue when I rebase and older branch: http://stackoverflow.com/questions/22412444/local-git-branch-has-diverged-from-origin-following-an-interactive-rebase .

**Comment by TheHollidayInn** [comment]
Woot. I think I figured it out.  I had to fetch upstream first. My b.

**Comment by TheHollidayInn** [comment]
Alright, so I added the ability to use both filters (they are andded together). Also, I have added tests. Let me know if we need anything else. 

I can probably start adding more challenge tests in a separate PR.

File: website/public/js/controllers/challengesCtrl.js [hunk 2]
```diff
@@ -256,9 +256,23 @@ habitrpg.controller("ChallengesCtrl", ['$rootScope','$scope', 'Shared', 'User',
...
+
+     //Currently, all filters are andded together
+
+     //First we must filter by selected groups
+     var showChallenge = $scope.search.group[chal.group._id];
+
+     //Membership filters: Show if user has selected either or has selcted Owned and is owner
+     var userIsOwner = chal.leader._id == User.user.id;
+     showChallenge = showChallenge && (typeof $scope.search._isOwner === 'undefined' || userIsOwner === $scope.search._isOwner );
+console.log($scope.search._isMember, chal._isMember)
```
Comment (crookedneighbor) [inline website/public/js/controllers/challengesCtrl.js]: Got an extraneous console log here.

File: test/spec/challengesCtrlSpec.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+      challenges = Challenges;
+
+      groups = Groups;
+    });
+  });
+
+  it('filters challenges', function() {
+
+   //There are for types of challenges
```
Comment (crookedneighbor) [inline test/spec/challengesCtrlSpec.js]: `s/for/four`

File: test/spec/challengesCtrlSpec.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+     rewards: [],
+     leader: user._id,
+     group: "test",
+     timestamp: +(new Date),
+     members: [],
+     official: false,
+     _isMember: false
+   });
+
+   //You not be the owner and a member
```
Comment (crookedneighbor) [inline test/spec/challengesCtrlSpec.js]: I think these comments should be in this format:

> // You are the owner and a member
> // You are the owner but not a member
> // You are not the owner but you are a member
> // You are not the owner or a member

**Comment by crookedneighbor** [comment]
Overall, loving this. You were very thorough with the spec tests. Thank you!

I put a new branch on upstream where I refactored your tests a bit. Here's the [specific commit](https://github.com/HabitRPG/habitrpg/commit/fcf8b556ef59250d2dbb167b8eb36b82a633b2cf?w=1), I've used the [ignore white space flag in the url](https://github.com/blog/967-github-secrets), so you should only see the real changes, not just a mass of indentations. 

Mostly, I split it up into many `it` blocks instead of one giant one. That required converting the current `it` into a `describe` and putting the setup of the different kind of challenges into a `beforeEach`.

That way the tests themselves could describe what they're doing, instead of comments above the tests, and you get this nice output:

screen shot 2015-04-23 at 8 17 42 am

You can:
 A) cherry pick the commit

``` bash
git fetch upstream
git cherry-pick fcf8b556ef59250d2dbb167b8eb36b82a633b2cf
```

B) Refactor a version for yourself so you can get a better understanding of how it works

C) Do nothing and I'll merge in my test refactor after your PR is merged in.

**Comment by TheHollidayInn** [comment]
I'm glad! I'm trying to add tests on all the changes I am submitting.

This is much cleaner! I wasn't sure what the better practice would be. I definitely like this method. Also, I wasn't aware you can use describe like that. I'll cherry pick in an hour or so to play around with it.

**Comment by crookedneighbor** [comment]
Yep. You can nest `describe`, but not `it`. 

Think of the describe blocks like contexts. The outer describe block is saying that this whole thing is the test for the challenges ctrl. The inner describe block is saying this is the test for the filter for the challenges controller. Makes organizing things much easier. 

Feel free to play around with the test and make changes that make sense to you.

**Comment by TheHollidayInn** [comment]
Awesome. I like this a lot. I'll apply it in the future for sure. I was thinking the tests should be smaller and easier to maintain.

I uploaded my changes and changed the language to follow the it - verb pattern, but fell free to simply merge your commit if you prefer that language.

**Comment by crookedneighbor** [comment]
Looks beautiful! 

The only issues I'm having with it are issues that are present in develop, so I'm not going to harp on having them fixed before this can be merged in. I'll put those in a separate ticket. Anyone else want to check this out before I merge it?

**Comment by TheHollidayInn** [comment]
Is this good to go?

**Comment by crookedneighbor** [comment]
@TheHollidayInn Sorry for the delay on this. 

Good work! Especially the tests. I've refactored the challenges filter a little to make it more readable. Nothing to do with your code really, it was pretty borked to begin with! https://github.com/HabitRPG/habitrpg/commit/6513527d8be7ba9476c9a01f97ef7251d3803700?w=1

I've passed the flag to ignore white space, so you don't have to pay attention to the indentation changes I made. 

Notated toward your next tier.

**Pull request merged**